### PR TITLE
Regression test implementation tested for single flavor

### PR DESCRIPTION
**Pull request by lilachmaliniak**
Regression test is implemented for multiple flavors but currently tested, rouphly, for single.
work in progress to fully support multiple flavours.
Known issue: 
chunklist checksums do not match between ground truth and test results.

### CODE REVIEW

File: lib/Adapters/RegressionAdapter/hlsStreamRegressionValidator.js [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    return this.loadResultsDatabase()
+        .then( function() {
+            that.logger.debug('[%s] regression tests validator successfully initialized.', that.entryId);
+        })
+        .catch (function(err) {
+            that.logger.error('[%s] failed to initialize regression tests validator. Error: %s, stack: %s', that.entryId, err.message, err.stack);
+        });
+}
+
+HlsStreamRegressionValidator.prototype.createDestinationDir = function() {
```
Comment (kalturaguy) [inline lib/Adapters/RegressionAdapter/hlsStreamRegressionValidator.js]: should +HlsStreamRegressionValidator.prototype.createDestinationDir  be in utils?

File: lib/Adapters/RegressionAdapter/validatorUtility.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+       else if (!valid) {
+           diff['result'] = 'regression test failed';
+       }
+       else {
+           diff['result'] = 'validation skipped. config \'override\' value=true';
+       }
+       var result_path = that.config.path;
+       result_path += '/' + that.entryId + '_regression_test_result.json';
+       // save diff
+       fs.writeFile(result_path, JSON.stringify(diff, null, 2), function (write_err) {
```
Comment (kalturaguy) [inline lib/Adapters/RegressionAdapter/validatorUtility.js]: use q-io writeFile instead

File: lib/Adapters/RegressionAdapter/hlsStreamRegressionValidator.js [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+            // todo: count files and delete old files so that total number including new one is up to history_count
+
+            if (that.override === true) {
+                that.logger.debug("[%s] creating new database for regression test results!", that.entryId);
+            }
+            else {
+                that.logger.debug('Going to open %s. Override=%s', that.fullpath, that.override);
+
+                fs.readFile(that.fullpath, 'utf8', function (err, data) {
```
Comment (kalturaguy) [inline lib/Adapters/RegressionAdapter/hlsStreamRegressionValidator.js]: use q-io readFile

File: lib/Adapters/RegressionAdapter/hlsStreamRegressionValidator.js [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+
+HlsStreamRegressionValidator.prototype.saveResultsToFile = function (reason) {
+
+    var that = this;
+    var deferred = Q.defer();
+
+    this.logger.info('[%s] saving regression results. Reason: %s', this.entryId, reason);
+
+    if (this.override || this.regression_results_db === undefined) {
+        fs.writeFile(this.fullpath, JSON.stringify(this.this_run_results, null, 2), function (err) {
```
Comment (kalturaguy) [inline lib/Adapters/RegressionAdapter/hlsStreamRegressionValidator.js]: use qio writefile

File: lib/Adapters/RegressionAdapter/RegressionAdapter.js [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+        var mediaServerPlaylistUrl = that.getFullLiveUrl(entryConfig.entryId, 'all');
+        var fullUrl = url.resolve(mediaServerPlaylistUrl, 'playlist.m3u8');
+
+        // read and parse the playlist and assemble playlist per flavor
+        return networkClient.read(fullUrl)
+            .then(function (content) {
+                that.initPlaylistPerFlavor(entryConfig.entryId, content);
+            })
+            .catch(function (error) {
+                that.logger.warn('Initialization of entryId failed with error %s. Failed to prepare playlist per flavor.', error);
```
Comment (kalturaguy) [inline lib/Adapters/RegressionAdapter/RegressionAdapter.js]: what about callstack

File: lib/Adapters/RegressionAdapter/RegressionAdapter.js [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+    }
+
+    return playlist;
+}
+
+TestHelper.prototype.getLastChunklistFileIndex = function (entryId, fullPath) {
+
+    var that = this;
+
+    if (!String.prototype.endsWith) {
```
Comment (kalturaguy) [inline lib/Adapters/RegressionAdapter/RegressionAdapter.js]: can be removed now with ES6

File: lib/Adapters/RegressionAdapter/RegressionAdapter.js [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+            }
+            position -= searchString.length;
+            var lastIndex = subjectString.indexOf(searchString, position);
+            return lastIndex !== -1 && lastIndex === position;
+        };
+    }
+
+    var deferred = Q.defer();
+
+    fs.readdir(fullPath, function (err, items) {
```
Comment (kalturaguy) [inline lib/Adapters/RegressionAdapter/RegressionAdapter.js]: there is already an implementation (glob?)

File: lib/Adapters/RegressionAdapter/hlsStreamRegressionValidator.js (hunk 3 above)
Comment (lilachmaliniak) [inline lib/Adapters/RegressionAdapter/hlsStreamRegressionValidator.js]: done, replaced with qio.write(...)

File: lib/Adapters/RegressionAdapter/hlsStreamRegressionValidator.js (hunk 1 above)
Comment (lilachmaliniak) [inline lib/Adapters/RegressionAdapter/hlsStreamRegressionValidator.js]: no need to add to utils. Replaced with call to q-io makeTree().

Comment (lilachmaliniak) [inline lib/Adapters/RegressionAdapter/hlsStreamRegressionValidator.js]: no need to add to utils. Replaced with call to q-io makeTree().

File: lib/Adapters/RegressionAdapter/hlsStreamRegressionValidator.js (hunk 3 above)
Comment (lilachmaliniak) [inline lib/Adapters/RegressionAdapter/hlsStreamRegressionValidator.js]: done.

File: lib/Adapters/RegressionAdapter/hlsStreamRegressionValidator.js (hunk 4 above)
Comment (lilachmaliniak) [inline lib/Adapters/RegressionAdapter/hlsStreamRegressionValidator.js]: done.

File: lib/Adapters/RegressionAdapter/RegressionAdapter.js (hunk 5 above)
Comment (lilachmaliniak) [inline lib/Adapters/RegressionAdapter/RegressionAdapter.js]: replaced in all places where error is printed to log with ErrorUtils.error2string(err).

File: lib/Adapters/RegressionAdapter/RegressionAdapter.js (hunk 6 above)
Comment (lilachmaliniak) [inline lib/Adapters/RegressionAdapter/RegressionAdapter.js]: removed.

**Comment by lilachmaliniak** [comment]
finished implementing all review comments.

File: lib/Adapters/RegressionAdapter/RegressionAdapter.js [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+
+TestHelper.prototype.initEntries = function() {
+
+    var that = this;
+    var root_mp4_path = config.get('rootFolderPath');
+    var p;
+    p = _.map(entriesInfo, function (entryConfig) {
+        // remove directory
+        var fullpath = root_mp4_path + '/' + entryConfig.entryId;
+        return fs_utils.cleanFolder(fullpath, 'obsolete').then(function () {
```
Comment (kalturaguy) [inline lib/Adapters/RegressionAdapter/RegressionAdapter.js]: cleanFolder of fs_utils is not really cleaning the folder, but it only renames it to another location (performance)
suggest you use https://www.npmjs.com/package/q-io#removetreepath

Comment (lilachmaliniak) [inline lib/Adapters/RegressionAdapter/RegressionAdapter.js]: done

File: lib/Adapters/AdapterFactory.js [hunk 9]
```diff
@@ -2,15 +2,22 @@
...
 var config = require('../../common/Configuration');
 
 var simulateStreams = config.get('simulateStreams');
 var regressionAdapterConfig = config.get('regressionAdapter');
+var command = process.argv.length > 2 ? process.argv[2] : undefined;
+
 
 exports.getAdapter = function() {
 
+    if (command != undefined) {
```
Comment (kalturaguy) [inline lib/Adapters/AdapterFactory.js]: this doesn't belong here  if you want to override configuration it should be in a code that handles config (maybe config.js...)

File: package.json [hunk 10]
```diff
@@ -2,10 +2,10 @@
...
   "version": "1.0.0",
   "repository": {
-      "type": "git",
-      "url": "https://github.com/kaltura/liveDVR"
-    },
-  "license": "",
+    "type": "git",
+    "url": "git+https://github.com/kaltura/liveDVR.git"
+  },
+  "license": "ISC",
```
Comment (kalturaguy) [inline package.json]: why ISC? is this aligned with other products in kaltura?

File: package.json [hunk 11]
```diff
@@ -22,22 +22,37 @@
...
-    "chai-as-promised":"4.1.1"
+    "chai-as-promised":"4.1.1",
+    "grunt-execute": "^0.2.2"
   },
   "dependencies": {
     "curry": "1.2.0",
     "forever": "0.15.2",
     "glob": "7.0.5",
     "log4js": "0.6.37",
+    "m3u8": "git://github.com/eladBenedict/node-m3u8.git#master",
```
Comment (kalturaguy) [inline package.json]: why did you re-add it?

File: package.json (hunk 10 above)
Comment (lilachmaliniak) [inline package.json]: I don't know. This property was added after automatic init to add grunt-execute.

Comment (lilachmaliniak) [inline package.json]: removed.

File: lib/Adapters/RegressionAdapter/hlsStreamRegressionValidator.js [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+const ErrorUtils = require('../../utils/error-utils');
+
+function checksum (str, algorithm, encoding) {
+    return crypto
+        .createHash(algorithm || 'md5')
+        .update(str, 'utf8')
+        .digest(encoding || 'hex')
+}
+
+let logheader1 = function(literals, ...values) {
```
Comment (kalturaguy) [inline lib/Adapters/RegressionAdapter/hlsStreamRegressionValidator.js]: this is duplicate code, the logger has already a build in feature of this, please use it

File: package.json [hunk 13]
```diff
@@ -36,8 +36,21 @@
...
     "underscore": "1.8.3",
     "v8-profiler": "5.6.5"
-  }
+  },
+  "description": "DVR implementation built on top of live streams",
+  "bugs": {
+    "url": "https://github.com/kaltura/liveDVR/issues"
+  },
+  "homepage": "https://github.com/kaltura/liveDVR#readme",
+  "main": "grunt-config.js",
```
Comment (kalturaguy) [inline package.json]: why the main entry is the grunt?

Comment (lilachmaliniak) [inline package.json]: This is a mistake. I change it to lib/App.js

File: common/config/config.json.template [hunk 14]
```diff
@@ -23,8 +23,29 @@
...
   "monitorPort": 4390,
+  "preserveOriginalHLS": false,
+  "recordingDurationMinutes": 5,
+  "overrideOldRecordedHLS": true,
   "regressionAdapter": {
-      "enable":false
+      "enable":false,
+      "analysis": false,
+      "analyzerRootPath" : "TBD",
+      "dataWarehouseRootFolderPath" : "instructions: put here root path to data warehouse (e.g. /var/tmp/regression/data_warehouse)",
```
Comment (kalturaguy) [inline common/config/config.json.template]: why not use ~/.....

File: package.json [hunk 15]
```diff
@@ -36,8 +38,22 @@
     "request": "^2.61.0",
     "shelljs": "0.7.0",
     "touch": "1.0.0",
-    "chunked-stream": "0.0.2",
     "underscore": "1.8.3",
-    "v8-profiler": "5.6.5"
-  }
+    "v8-profiler": "5.6.5",
+    "child_process": "1.0.2"
+  },
```
Comment (kalturaguy) [inline package.json]: you removed chunked-stream, but it's still used

File: package.json [hunk 16]
```diff
@@ -36,8 +38,22 @@
     "request": "^2.61.0",
     "shelljs": "0.7.0",
     "touch": "1.0.0",
-    "chunked-stream": "0.0.2",
     "underscore": "1.8.3",
-    "v8-profiler": "5.6.5"
-  }
+    "v8-profiler": "5.6.5",
+    "child_process": "1.0.2"
```
Comment (kalturaguy) [inline package.json]: please don't use this package, node.js has already all functionality you need

File: lib/utils/general-utils.js [hunk 17]
```diff
@@ -0,0 +1,30 @@
+/**
```
Comment (kalturaguy) [inline lib/utils/general-utils.js]: don't think this belong here, this is used only by ControllerWrapper, so it should be there

File: lib/Controller.js [hunk 18]
```diff
@@ -1,7 +1,8 @@
 /**
  * Created by elad.benedict on 8/26/2015.
  */
-
+// parse command line args and update config obj
+var regressionConfig = require('./Adapters/RegressionAdapter/RegressionConfig.js').getConfig();
```
Comment (kalturaguy) [inline lib/Controller.js]: this shouldn't belong here, the Controller shouldn't know what regressionAdapter is

File: lib/Controller.js [hunk 19]
```diff
@@ -226,16 +235,33 @@ module.exports = (function(){
...
+                // Todo: check if a bug --> stop() doesn't work as expected
+                // that.stop();
+            }
+        });
+        adapter = adapter.getAdapter(that);
+
         logger.warn("Starting liveProxy pid: %s, wowza: '%s', API server: '%s', prefix='%s'",process.pid,config.get("mediaServer").hostname,config.get('backendClient').serviceUrl,that.prefix);
         logger.warn("Log files %s",config.get('logFileName'));
         return Q.fcall(function() {
+            if (regressionConfig.command && regressionConfig.command.record ||
```
Comment (kalturaguy) [inline lib/Controller.js]: please move it from here... (same comment from above) you already have Controller wrapper

File: lib/Adapters/TestAdapter.js [hunk 20]
```diff
@@ -10,6 +10,7 @@ var networkClient = require('./../NetworkClientFactory').getNetworkClient();
 var url=require('url');
```
Comment (kalturaguy) [inline lib/Adapters/TestAdapter.js]: why did you need to change this file?

File: lib/Adapters/RegressionAdapter/hlsStreamRegressionValidator.js [hunk 21]
```diff
@@ -0,0 +1,296 @@
...
+
+    var that = this;
+    var deferred  = Q.defer();
+
+    data.errors[data.flavor] = {errors: {}};
+    var flavor_header = logheader1 `${this.entryId}${data.flavor}${data.index}`;
+
+    let url = util.format("http://localhost:8080/kLive/smil:%s_all.smil/%s/%s", this.entryId, data.flavor, item.properties.uri);
+
+    networkClient.read({url: url, timeout: 10000})
```
Comment (kalturaguy) [inline lib/Adapters/RegressionAdapter/hlsStreamRegressionValidator.js]: you are downloading the entire file just to know the length....
i suggest you do a HEAD http request which will return the length and not the content

File: lib/Adapters/RegressionAdapter/StreamRegressionValidatorFactory.js [hunk 22]
```diff
@@ -0,0 +1,23 @@
...
+var hlsStreamRegressionValidator = require('./hlsStreamRegressionValidator');
+var hlsAnalysisStreamRegressionValidator = require('./hlsAnalysisStreamRegressionValidaotor');
+var analysis = config.get('regressionAdapter').analysis;
+
+var getValidator = function(entryConfig, flavors, last_chunklist_index) {
+
+    if (analysis) {
+        return new hlsAnalysisStreamRegressionValidator(entryConfig, flavors, last_chunklist_index);
+    } else {
+        return new hlsStreamRegressionValidator(entryConfig, flavors, last_chunklist_index);
```
Comment (kalturaguy) [inline lib/Adapters/RegressionAdapter/StreamRegressionValidatorFactory.js]: the name is not good... maybe HlsChecksumRegressionValidator ?

File: common/config/config.json.template (hunk 14 above)
Comment (lilachmaliniak) [inline common/config/config.json.template]: replaced with "~/live-monitor/"

File: package.json (hunk 15 above)
Comment (lilachmaliniak) [inline package.json]: added chunked-stream and removed child_process.

File: package.json (hunk 16 above)
Comment (lilachmaliniak) [inline package.json]: ok - code will change

File: lib/utils/general-utils.js (hunk 17 above)
Comment (lilachmaliniak) [inline lib/utils/general-utils.js]: removed

File: package.json (hunk 16 above)
Comment (lilachmaliniak) [inline package.json]: using node original child_process.

File: lib/Controller.js (hunk 18 above)
Comment (lilachmaliniak) [inline lib/Controller.js]: moved to AdapterFactory.js. Controller code cleaned from unit tests related code.

File: lib/Adapters/TestAdapter.js (hunk 20 above)
Comment (lilachmaliniak) [inline lib/Adapters/TestAdapter.js]: 1) support multiple different recording of same entry, based on override flag.
2) fixed the naming convention of recorded stream output directory to avoid using spaces in name.
3) added timer to limit recording time based on configuration.

File: lib/Adapters/RegressionAdapter/hlsStreamRegressionValidator.js (hunk 21 above)
Comment (lilachmaliniak) [inline lib/Adapters/RegressionAdapter/hlsStreamRegressionValidator.js]: done. Now request options obj include method: HEAD.

File: lib/Adapters/RegressionAdapter/StreamRegressionValidatorFactory.js (hunk 22 above)
Comment (lilachmaliniak) [inline lib/Adapters/RegressionAdapter/StreamRegressionValidatorFactory.js]: replaced. did more renaming:
hlsStreamRegressionValidator --> hlsChunklistRegressionValidator  StreamRegressionValidatorFactory --> RegressionValidatorFactory 
StreamRegressionValidatorBase --> RegressionValidatorBase  hlsAnalysisStreamRegressionValidator --> hlsAnalysisRegressionValidator

File: lib/Adapters/RegressionAdapter/hlsChecksumRegressionValidator.js [hunk 23]
```diff
@@ -0,0 +1,300 @@
...
+    //let url = util.format("http://localhost:8080/kLive/smil:%s_all.smil/%s/%s", this.entryId, data.flavor, item.properties.uri);
+    let options = {
+        method: 'HEAD',
+        url: `http://localhost:8080/kLive/smil:${this.entryId}_all.smil/${data.flavor}/${item.properties.uri}`,
+       
+        timeout: 10000};
+    networkClient.read(options)
+        .then(function (content) {
+            // add to results new segment
+            var ts_id = item.properties.uri.split('.')[0];
```
Comment (kalturaguy) [inline lib/Adapters/RegressionAdapter/hlsChecksumRegressionValidator.js]: if you are doing head request then what is content now?  content.length will probably be something else... you need the header 'content-length'

File: lib/Adapters/RegressionAdapter/hlsChecksumRegressionValidator.js [hunk 24]
```diff
@@ -0,0 +1,300 @@
...
+            if (!_.has(that.this_run_results[data.flavor].ts[ts_id])) {
+                that.this_run_results[data.flavor].ts[ts_id] = content.length;
+                deferred.resolve();
+            } else if (that.this_run_results[flavorId].ts[ts_id] != content.length) {
+                that.logger.debug('%s %s validation failed', flavor_header, ts_id);
```
Comment (kalturaguy) [inline lib/Adapters/RegressionAdapter/hlsChecksumRegressionValidator.js]: instead of  passing the flavor_header on each log line, it's better if you have created a new logger with this as prefix, see example in:

https://github.com/kaltura/liveDVR/blob/1.0/lib/entry/FlavorDownloader.js#L55

File: lib/Adapters/RegressionAdapter/hlsChecksumRegressionValidator.js (hunk 23 above)
Comment (lilachmaliniak) [inline lib/Adapters/RegressionAdapter/hlsChecksumRegressionValidator.js]: done. The length is taken from content-length header filed. It required also update of NetworkClient to retune respose.headers instead of body, when method: HEAD is part of request.

File: gruntfile.js [hunk 25]
```diff
@@ -11,9 +11,13 @@ module.exports = function (grunt) {
...
 
+    grunt.loadNpmTasks('grunt-execute');
+
     grunt.loadTasks('./lib/grunt/tasks');
     grunt.loadTasks('./lib/grunt/config');
 
+
     // Default task(s).
-    grunt.registerTask('default', ['jshint:dev','unit-test:dev', 'component-test:dev', 'mocha_istanbul:dev']);
+   // grunt.registerTask('default', ['jshint:dev', 'execute:regression_test', 'unit-test:dev', 'component-test:dev', 'mocha_istanbul:dev']);
```
Comment (kalturaguy) [inline gruntfile.js]: is now the unit-test disabled?

File: common/config/config.json.template [hunk 26]
```diff
@@ -23,8 +23,29 @@
   "logFileName" : "~/dvr/kLiveController.log",
   "logTimeZoneOffset" : 300,
   "monitorPort": 4390,
+  "preserveOriginalHLS": false,
+  "recordingDurationMinutes": 5,
```
Comment (kalturaguy) [inline common/config/config.json.template]: can you please move that to regessionAdapter or such?  it will be confusing for the future when we will do the real live to vod recording in the live controller

File: lib/NetworkClient.js [hunk 27]
```diff
@@ -19,7 +19,11 @@ function readRequest(requestData)
             return;
         }
         if (!error && response.statusCode === 200) {
-            d.resolve(body);
+            if (!requestData.method || requestData.method != 'HEAD') {
```
Comment (kalturaguy) [inline lib/NetworkClient.js]: don't like this much.... 
maybe add resolveWithFullResponse to the requestData and if this true then return
d.resolve ({body,response.headers})
(so we can reuse this)

File: lib/Controller.js [hunk 28]
```diff
@@ -253,14 +259,18 @@ module.exports = (function(){
...
-
+    
     Controller.prototype.getStat = function() {
         return {
             "handledEntries": this.handledEntries,
             "phantomEntries": this.phantomEntries
         }
     };
 
+    Controller.prototype.getAdapter = function() {
```
Comment (kalturaguy) [inline lib/Controller.js]: why exposing adapter with the controller?  you can call 
var adapter = require('./Adapters/AdapterFactory.js').getAdapter();
anywhere in code...

File: lib/Controller.js [hunk 29]
```diff
@@ -15,8 +16,10 @@ var Q = require('q');
 var qio = require('q-io/fs');
 var path = require('path');
 var util = require('util');
-var adapter = require('./Adapters/AdapterFactory.js').getAdapter();
+
 var MonitorServer=require('./MonitorServer.js');
+const events = require('events');
+
```
Comment (kalturaguy) [inline lib/Controller.js]: nothing in this file really change, so for each merge revert all space changes so this file will not be in the pull reuqest

File: lib/Controller.js (hunk 28 above)
Comment (lilachmaliniak) [inline lib/Controller.js]: removed getAdapter from Controller.js and updated AdapterFactory to support only one timer adapter instantitiation.

File: lib/Controller.js (hunk 29 above)
Comment (lilachmaliniak) [inline lib/Controller.js]: Guy, there is one change that is required, unless I add the line to App.js...
var adapter = require('./Adapters/AdapterFactory.js').getAdapter(); moved to the top of Controller.js so the configuration updates will take place before any module get config.

Comment (lilachmaliniak) [inline lib/Controller.js]: Guy,
there is one change that is required, unless I add the line to App.js...
var adapter = require('./Adapters/AdapterFactory.js').getAdapter();
moved to the top of Controller.js so the configuration updates will take place before any module get config.
